Basics 1 (#31)
mlr-org/mlr3book: bookdown/01-mlr3-basics.Rmd

# mlr3 Basics {#basics}

This chapter will teach you the essential building blocks, R6 classes, and operations of `r mlr_pkg("mlr3")`.

> This includes creating supervised machine learning tasks like classification and regression, training models and getting prediction on new data, and evaluating and comparison of different models through cross-validation and benchmarking.

Reviewer: ```suggestion
This includes creating supervised machine learning tasks like classification and regression, training models and getting prediction on new data, and evaluating and comparing different models through cross-validation and benchmarking.
```

> A typical machine learning workflow looks like this:

Reviewer: Should we make a global decision how we want ML to be written? I.e. either

- machine-learning
- machine learning

Reviewer 2: in english you do not use hyphens for this.... so pls "machine learning" or even ML

```{r 01-mlr3-basics-001, echo = FALSE}
knitr::include_graphics("images/ml_abstraction.png")
```

The data, which `r mlr_pkg("mlr3")` encapsulates in tasks, is split into non-overlapping train and test sets to be able to evaluate models objectively &mdash; we are interested in models that generalize to new data rather than just memorizing the training data.
The training data is given to a machine learning algorithm, called a learner in `r mlr_pkg("mlr3")`, that uses it to build a model of how the features of the data relate to the target values.
This model is then used to produce predictions on the test data, which are compared to the ground truth values to assess the quality of the model.

> `r mlr_pkg("mlr3")` offers a number of different measures to quantify this quality; usually a numeric score.

Reviewer: > usually a numeric score

If we mention it explicitly, should we also give an example? Which measures are not a numeric score?

Author: In particular I was thinking of confusion matrices. Yes, they're not really measures as such, but they fall into the same category. That said, I don't feel strongly about this if you want to change it.

Reviewer 2: i would keep the matrices and measures separate. both should be covered. but in terms of mlr3 a measure always returns a scalar 1-number score

This process may be repeated several times, each time resampling different train and test sets from the original data set.

> Multiple resampling iterations allow to get a better generalization performance estimate for a particular type of model by quantifying its performance on different data.

Reviewer: ```suggestion
Multiple resampling iterations allow to get a generalized performance estimate for a particular type of model by quantifying its performance on different data.
```

Author: Don't agree with this change -- "generalized performance estimate" qualifies "performance estimate", but we're not doing anything to take an existing performance estimate and make it somehow more general. "Generalization performance estimate" is a different kind of performance estimate, i.e. a general one.

Reviewer 2: lars is correct

The `r mlr_pkg("mlr3")` package provides R6 classes for the essential building blocks of this machine learning workflow:

* A [task](#tasks) encapsulates the data along with additional information, such as what the prediction target is.
* A [learners](#learners) encapsulates one of R's many machine learning algorithms and allows to train models and make predictions.
  Most learners have hyperparameters that affect their operation.
* A [measure](#measures) computes a numeric score based on predicted and ground-truth values and their difference.
* A [resampling](#resampling) specifies a series of train and test sets.

In many cases, this simple workflow is not sufficient to deal with real-world data, which may require normalization, imputation of missing values, or feature selection.

> We will cover more complex workflows that allow to do this and much more later, but restrict ourselves to simple workflows like the one above here for the sake of clarity.

Reviewer: ```suggestion
We will cover more complex workflows that allow to do this and even more later in the book.
For now, we restrict ourselves to simple workflows like the one above for the sake of clarity.
```
